### ISSUE TITLE
[RFC] Please add an example how to use Microsoft Azure Attestation(MAA) for a remote attestation flow

### ISSUE THREAD

**Issue by agowdamsft** [occlum/occlum#797]
- Feature Name: Please add a demo/sample that shows how to do remote attestation using MAA
- Start Date: 01/21/2022

# Summary
[summary]: Remote attestation is extremely important for establishing trust with the apps. MAA is a popular choice for customers deploying on Azure.

Request is to extend the current remote attestation sample and add how a 3rd party/client was able to request the SGX quote of the occlum app and get it attested through MAA. More details and samples on MAA are here https://docs.microsoft.com/en-in/azure/attestation/overview

# Motivation
[motivation]: Direct Azure customer requests

Why are we doing this? What use cases does it support? What is the expected outcome?

# Guide-level explanation
[guide-level-explanation]: #guide-level-explanation

Explain the proposal as if it was already included in the Occlum and you were teaching it to another Occlum user. That generally means:

-extending the current ability

**Comment by qzheng527** [comment]
@agowdamsft Sorry for the late reply cause of the Chinese New Year holiday.
We are happy to extend Occlum usage into Azure.
I just checked the MAA overview, looks that the core part to add Occlum based MAA demo is to provide similar `genquote_host` and `genquote_enclave` implementation. I am going to follow up on this part.
One question for you, in the end how can we contribute this demo to Azure? 
Thanks.

**Comment by agowdamsft** [comment]
@qzheng527 once the sample is ready here I am happy to work with you for azure samples using a real customer scenario here https://github.com/Azure-Samples/confidential-container-samples

**Comment by qzheng527** [comment]
@agowdamsft Great, Occlum is happy to be part of the Aruze confidential computing.
One more question about the MAA json format.
Evert part is clear from the [`source`](https://github.com/Azure-Samples/microsoft-azure-attestation/blob/master/intel.sdk.attest.sample/genquotes/host/host.cpp#L162) except the `EnclaveHeldDataHex`.
Is it compulsory that the `sgx_report_data` needs put the raw data after [`sha256`](https://github.com/Azure-Samples/microsoft-azure-attestation/blob/master/intel.sdk.attest.sample/genquotes/host/host.cpp#L100) but the json file requires the [`raw data`](https://github.com/Azure-Samples/microsoft-azure-attestation/blob/master/intel.sdk.attest.sample/genquotes/host/host.cpp#L168) only ?

**Comment by olkroshk** [comment]
Hi, @qzheng527, `QuoteHex` and `EnclaveHeldDataHex` are the minimal required parameters of the JSON file that should be sent to MAA as a part of the attest calls. I.e. JSON at the very minimum should contain the SHA256 hash and the actual content (https://github.com/Azure-Samples/microsoft-azure-attestation/blob/master/intel.sdk.attest.sample/genquotes/host/host.cpp#L167-L168).
 
The rest of the parameters were added to perform additional validations, see [OE Sample](https://github.com/Azure-Samples/microsoft-azure-attestation/blob/master/sgx.attest.sample/validatequotes.core/EnclaveInfo.cs#L24) and, similarly, [Intel SGX Sample](https://github.com/Azure-Samples/microsoft-azure-attestation/blob/master/intel.sdk.attest.sample/validatequotes.core/EnclaveInfo.cs#L26).

Please let me know if this answers your question.

Thank you,
Olga

**Comment by qzheng527** [comment]
@agowdamsft @olkroshk I submitted a new PR for the MAA demo.
https://github.com/occlum/occlum/pull/807

The full Microsoft Azure Attestation flow includes generating a quote in an SGX enclave and then get it validated by the Microsoft Azure Attestation (MAA) service.

1. Build an SGX enclave
2. Launch an SGX enclave and get SGX quote
3. Persist SGX quote and Enclave Held Data (EHD) to JSON file
4. Call Azure Attestation for validation
5. Output validation results

This demo only covers the first three steps.
Please refer to the [`README`](https://github.com/occlum/occlum/pull/807/commits/0ffaec001d74634905dc01d0b7ef28ab2358b78c#diff-2d3e0fe8b1f0bd1a328241cde1fc1c9079e53dce897c2927cb0879ea29f9cef6) for how to use.

Just remember to try the demo in the docker image `occlum/occlum:0.26.4-ubuntu18.04`.
Please let me know if any missing in the PR.

**Comment by qzheng527** [comment]
@agowdamsft @kroshkina-msft Any feedback about the Occlum MAA demo?

**Comment by kroshkina-msft** [comment]
@qzheng527, we are looking at the demo. I will run a test and get back to you.

`+` @gkostal

**Comment by kroshkina-ms** [comment]
@qzheng527 
The code looks good to me. 
However, I tried to run the demo a few times and I get the following error:
```
Built the Occlum image and enclave successfully
occlum run to generate quote in maa json format
gen_maa_json: error while loading shared libraries: libocclum_dcap.so.0.1.0: cannot open shared object file: No such file or directory
```
Please let me know if this is a known issue.

**Comment by qzheng527** [comment]
@kroshkina-ms The error was due to the soname adding for the occlum dcap lib.
So the quick fix could be as below.
```
--- a/demos/remote_attestation/maa/maa.yaml
+++ b/demos/remote_attestation/maa/maa.yaml
@@ -11,6 +11,6 @@ targets:
   - target: /opt/occlum/glibc/lib
     copy:
       - files:
-        - /opt/occlum/toolchains/dcap_lib/glibc/libocclum_dcap.so
+        - /opt/occlum/toolchains/dcap_lib/glibc/libocclum_dcap.so.0.1.0
```
I will update the PR as well.

**Comment by kroshkina-ms** [comment]
@qzheng527
Great, thank you!
Do you know if there is an easy way to add a symbolic link to `/opt/occlum/toolchains/dcap_lib/glibc/libocclum_dcap.so` to avoid potential issues with the versions?

Here is another error (after the fix of specifying the correct shared object is applied):
```
Built the Occlum image and enclave successfully
occlum run to generate quote in maa json format
[get_platform_quote_cert_data ../qe_logic.cpp:347] Error returned from the p_sgx_get_quote_config API. 0xe046
thread '<unnamed>' panicked at 'IOCTRL IOCTL_GET_DCAP_QUOTE_SIZE failed', src/occlum_dcap.rs:70:13
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
fatal runtime error: failed to initiate panic, error 5
```

**Comment by qzheng527** [comment]
@kroshkina-ms I already fixed the link issue in the [`branch`](https://github.com/qzheng527/occlum/tree/maa) and also the [`PR`](https://github.com/occlum/occlum/pull/814). I will rebase this on the latest Occlum version (0.27.0 for now) and let you know.

For the new error, probably it was due to the PCCS_URL settings.
Did you modify the PCCS_URL in the running container (occlum/occlum:0.26.4-ubuntu18.04) file /etc/sgx_default_qcnl.conf ?
I didn't see this step in your environment [`setup`](https://github.com/kroshkina-ms/shared-notes/blob/main/maa-occlum-demo-trial.md). And please also make sure the running container could access the PCCS service you created. I didn't see "--network host" parameter (or other solution you applied in your env)  when you started the occlum container.

**Comment by kroshkina-ms** [comment]
@qzheng527 
Thank you so much for the quick replies!
I did update the  `/etc/sgx_default_qcnl.conf` file.

However, I don't have a way to run an SGX-enabled host with Ubuntu 20.04. Is the demo supposed to work with 18.04 host and docker image `occlum/occlum:0.27.0-ubuntu20.04`?

After trying to setup and run the demo, I would recommend updating the `README.md` with steps that a non-prepared customer could easily use it.

**Comment by qzheng527** [comment]
@kroshkina-ms Yes, it could work with SGX-enabled host with Ubuntu 18.04 and docker image occlum/occlum:0.27.0-ubuntu20.04.
As for the README update, sure, I will update it. Thanks for your suggest.

**Comment by kroshkina-ms** [comment]
@qzheng527 

Question for you about your comment:
> 1. Build an SGX enclave
> 2. Launch an SGX enclave and get SGX quote
> 3. Persist SGX quote and Enclave Held Data (EHD) to JSON file
> 4. Call Azure Attestation for validation
> 5. Output validation results
This demo only covers the first three steps.

Are there any specific plans about steps 4 and 5? 
If not, the easiest way to complete the full attestation loop would be to use the validate quote sample from here: https://github.com/Azure-Samples/microsoft-azure-attestation/tree/master/intel.sdk.attest.sample/validatequotes.core 

`+` @mssindhurid to the thread

**Comment by qzheng527** [comment]
@kroshkina-ms Did you successfully generate the MAA json files?
As for the steps 4 and 5, they are beyond Occlum and could run on any platform.
And you are right about the easiest validation way.

> If not, the easiest way to complete the full attestation loop would be to use the validate quote sample from here: https://github.com/Azure-Samples/microsoft-azure-attestation/tree/master/intel.sdk.attest.sample/validatequotes.core

Could you follow up in your side for steps 4 and 5 since I don't have avaliable aruze account right now.

**Comment by qzheng527** [comment]
@kroshkina-ms 
`Error returned from the p_sgx_get_quote_config API. 0xe019` usually means the network issue.
May I see the content of the /etc/sgx_default_qcnl.conf **in your running Occlum container**?
I didn't see this file modification in your steps.
It is my setting to a locally setup PCCS service (similar to what you did just except that I set up the PCCS service in another container).
```
# PCCS server address
PCCS_URL=https://localhost:8081/sgx/certification/v3/

# To accept insecure HTTPS certificate, set this option to FALSE
USE_SECURE_CERT=FALSE

# You can use Intel PCS to get quote verification collateral
#COLLATERAL_SERVICE=https://api.trustedservices.intel.com/sgx/certification/v3/

# If you use PCCS service to get quote verification collateral, you can specify which API version is to be used
# The legacy 3.0 API will return CRLs in HEX encoded DER format, while the new 3.1 API will return raw DER format
#PCCS_API_VERSION=3.1

# Maximum retry times for QCNL. If RETRY is not defined or set to 0, no retry will be performed.
# It will first wait one second and then for all forthcoming retries it will double the waiting time
# By using RETRY_DELAY you disable this exponential backoff algorithm
#RETRY_TIMES=6

# Sleep this amount of seconds before each retry when a transfer has failed with a transient error
#RETRY_DELAY=10
```
Generally speaking, you just need modify `PCCS_URL` and `USE_SECURE_CERT` according to your env.

By the way, is there available PCCS service in Aruze? If yes, you don't need set up a PCCS service by yourself.
I am just wondering for the existed MAA demo  [`Intel SDK`](https://github.com/Azure-Samples/microsoft-azure-attestation/blob/master/intel.sdk.attest.sample),  is setting up PCCS service necessary step as well?

**Comment by bodzhang** [comment]
The sample definitely helps the customers to understand how to interact with MAA. It will be even better if the attestation and secret provisioning can be made "transparent" to the developer, only require the developer to configure the Attestation/KMS service to be used, for example, similar to [Mystikos' design](https://github.com/deislabs/mystikos/blob/main/doc/design/secret-provisioning.md).

**Comment by qzheng527** [comment]
@bodzhang Occlum has one suggested [`init-ra`](https://github.com/qzheng527/occlum/tree/enclave-cc/demos/remote_attestation/init_ra_flow) way to make the scecret management "transparent" to the developer based on the Occlum unique two-stages boot flow, DCAP and GRPC-RATLS.

**Comment by kroshkina-ms** [comment]
@qzheng527 could you please confirm which Intel SGX API is used to produce the quote that is written into the output JSON files?

**Comment by qzheng527** [comment]
@kroshkina-ms Occlum is a LibOS, application running in the Occlum, such as the MAA demo, doesn't directly call the Intel SGX APIs. Back to your question, the Quote is generated by Occlum provided [DCAP APIs/Libs](https://github.com/occlum/occlum/blob/master/docs/remote_attestation.md). These APIs of course eventually call the Intel SGX APIs through [OCALLs](https://github.com/occlum/occlum/blob/master/src/pal/src/ocalls/attestation.c#L100).

May I know is there any problem met in verifying the JSON files through MAA service?

**Comment by kroshkina-ms** [comment]
@qzheng527 

Yes, MAA attestation fails to attest generated quotes with:

```
Unhandled exception. System.AggregateException: One or more errors occurred. (Attestation Enclave operation failed with 65534: oe_verify_evidence failed with quote collateral error
0x3 - OE_INVALID_PARAMETER: 0xfffe: 'The specified quote was invalid due to an error in the quote collateral.;
RETURN_IF_ERROR('sgxQuote->Parse(serializedQuote)') failed with 0xfffe: 'The specified quote was invalid due to an error in the quote collateral.'
RETURN_IF_ERROR('SgxQuote::Deserialize( _crypto, inputQuote, ignoreQuoteSignatureMismatch, _isOpenEnclaveProcessor, sgxQuote)') failed with 0xfffe: 'The specified quote was invalid
due to an error in the quote collateral.'
```

That is why I was looking which for exact Intel SGX API is called because `occlum` provides both remote and local report and quote generation respectively. So, I am wondering if the demo might be using an API unsupported by MAA. For instance, https://github.com/occlum/occlum/blob/8db6a55696fd2cd38d966f9afcac09ed76b5fc8f/src/libos/src/util/sgx/sgx_report.rs#L14 

However, if the demo calls the API you specified (https://github.com/occlum/occlum/blob/master/src/pal/src/ocalls/attestation.c#L106), it should work.

**Comment by qzheng527** [comment]
@kroshkina-ms The generated quote should be fine because many of our other projects are using the same way to successfully generate/verify the quote (not MAA way). Perhaps you can double check the MAA json generation in the demo, if it is expected format., especially the "Type". I just used "3"  to make it different from the "2" for OE_ENCLAVE_TYPE_SGX.
https://github.com/occlum/occlum/blob/463d87d672336e45f06ca9100719b45224a93f05/demos/remote_attestation/maa/gen_quote/gen_maa_json.c#L98

**Comment by qzheng527** [comment]
@kroshkina-ms Any update from your side?

**Comment by kroshkina-ms** [comment]
we are working on the debugging and will get back to you shortly

**Comment by kroshkina-ms** [comment]
@qzheng527 we have done an investigation, and the quotes produces by occlum demo fail MAA attestation. 

Although the quote validation passes, the supplied runtime data digest does not match runtime data digest in quote.
In other words, the report data from the quote does not match the hash of the enclave held data.

Could you please verify what exactly is happening withing the `dcap_generate_quote` call ( https://github.com/occlum/occlum/blob/463d87d672336e45f06ca9100719b45224a93f05/demos/remote_attestation/maa/gen_quote/gen_maa_json.c#L76)?
 
Once again, MAA performs several attestation steps, including the `oe_verify_evidence` OpenEnclave API verification step, and additionally the verification of the hash of the enclave held data comparing to the report data from the quote. And the latter is the one that fails.

Please let us know if more information is needed. 

`+` @rrathna who is also working on the integration.

**Comment by qzheng527** [comment]
@kroshkina-ms Great, very useful information!
> In other words, the report data from the quote does not match the hash of the enclave held data.
Double check the code, probably it is because the struct `hash` doesn;t got initialized to 0 by default.
I Updated the PR with one more commit:

- Initialize the `hash` struct members to all 0.
- Do compare of the supplied "report data" (`hash`) and the report data from the generated quote to make sure they are same.

Please  try the latest [`maa-update`](https://github.com/qzheng527/occlum/tree/maa-update).

As for your question about the details of `dcap_generate_quote`.
In general eventually it is just calling the below SGX APIs through ocall.
```
sgx_qe_get_target_info
sgx_qe_get_quote_size
sgx_qe_get_quote
```

**Comment by kroshkina-ms** [comment]
@qzheng527 makes sense and I can confirm that the new code passes attestation, thank you. Could you update your PR please

**Comment by qzheng527** [comment]
@kroshkina-ms  Excellent, I created a new PR [`maa pr`](https://github.com/occlum/occlum/pull/895) including the fix.  And please let me know what is the next step? Any help needed from us?

**Comment by qzheng527** [comment]
It is already done. Three different azure MAA demos could be found [`here`](https://github.com/occlum/occlum/tree/master/demos/remote_attestation/azure_attestation)

**Comment by agowdamsft** [comment]
closing. thanks for the work here.
